### PR TITLE
cluster_resolver: wait until the configs have been propagated to all the children

### PR DESCRIPTION
**Pull request by aranjans**
### What does this PR do?
This PR ensures that every time configuration update is triggered, it waits for child policy configuration update inline.

### Approach:
Added a channel to `ccUpdate` of which that channel will be closed after child policies are updated. Same is verified in the test by having a `ClientConnUpdateHook()` which is called after the above mentioned channel is closed, which ensures child policy was updated synchronously.

RELEASE NOTES:

- cluster_resolver: wait for child policy configuration to be updated inline

### CODE REVIEW

File: xds/internal/balancer/clusterresolver/clusterresolver.go [hunk 1]
```diff
@@ -51,6 +51,9 @@ var (
 	newChildBalancer  = func(bb balancer.Builder, cc balancer.ClientConn, o balancer.BuildOptions) balancer.Balancer {
 		return bb.Build(cc, o)
 	}
+	// Below function is no-op in actual code, but can be overridden in
+	// tests to give tests visibility into exactly when certain events happen.
+	ClientConnUpdateHook = func() {}
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/clusterresolver.go:56]: // This function is a no-op in production code but can be overridden in tests 
// to track the timing of specific events for test visibility.

File: xds/internal/balancer/clusterresolver/clusterresolver.go [hunk 2]
```diff
@@ -306,6 +310,11 @@ func (b *clusterResolverBalancer) run() {
 			switch update := u.(type) {
 			case *ccUpdate:
 				b.handleClientConnUpdate(update)
+				if update.done != nil {
+					// Close the channel to convey to the consumers of updateCh
+					// that child policies config are updated inline.
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/clusterresolver.go:315]: // Close the channel to signal to consumers of updateCh
// that the child policies' configuration has been updated.

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go [hunk 3]
```diff
@@ -445,3 +446,64 @@ func (s) TestOutlierDetectionConfigPropagationToChildPolicy(t *testing.T) {
 		t.Fatalf("Timeout when waiting for child policy to receive its configuration")
 	}
 }
+
+// Test verifies that LB waits for child policy configuration update
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go:450]: // TestChildPolicyConfigUpdateBlocking ensures that the load balancer 
// blocks and waits for the child policy configuration to be updated
// in response to configuration changes.

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go [hunk 4]
```diff
@@ -445,3 +446,64 @@ func (s) TestOutlierDetectionConfigPropagationToChildPolicy(t *testing.T) {
...
+// inline on receipt of configuration update.
+func (s) TestChildPoliciesConfigUpdatedInline(t *testing.T) {
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go:452]: name suggestion: TestChildPolicyConfigUpdateBlocking

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go [hunk 5]
```diff
@@ -445,3 +446,64 @@ func (s) TestOutlierDetectionConfigPropagationToChildPolicy(t *testing.T) {
...
+	// Override the newConfigHook to ensure picker was updated.
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go:453]: // Override the ClientConnUpdateHook to track when the picker is updated.

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go [hunk 6]
```diff
@@ -445,3 +446,64 @@ func (s) TestOutlierDetectionConfigPropagationToChildPolicy(t *testing.T) {
...
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{e2e.DefaultCluster(clusterName, edsServiceName, e2e.SecurityLevelNone)},
+		Endpoints:      []*v3endpointpb.ClusterLoadAssignment{e2e.DefaultEndpoint(edsServiceName, "localhost", []uint32{testutils.ParsePort(t, server.Address)})},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go:489]: nit: "Failed to update resources on management server: %v", err

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go [hunk 7]
```diff
@@ -445,3 +446,64 @@ func (s) TestOutlierDetectionConfigPropagationToChildPolicy(t *testing.T) {
...
+	// resolver, and dial the test backends.
+	cc, cleanup := setupAndDial(t, bootstrapContents)
+	defer cleanup()
+
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		t.Fatalf("EmptyCall() failed: %v", err)
+	}
+
+	// Close the listener and ensure that the config is updated.
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go:502]: // Ensure that the child policy config update is processed.

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go [hunk 8]
```diff
@@ -445,3 +446,64 @@ func (s) TestOutlierDetectionConfigPropagationToChildPolicy(t *testing.T) {
...
+	lis.Close()
+	select {
+	case <-clientConnUpdateDone:
```
Comment (purnesh42H) [inline xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go:505]: suggestion: can add a comment in success block `// Success: config update was processed within the expected time.`

Comment (aranjans) [reply xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go]: Updated

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go (hunk 7 above)
Comment (aranjans) [reply xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go]: Updated

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go (hunk 6 above)
Comment (aranjans) [reply xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go]: Updated

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go (hunk 5 above)
Comment (aranjans) [reply xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go]: Updated

File: xds/internal/balancer/clusterresolver/clusterresolver.go (hunk 2 above)
Comment (aranjans) [reply xds/internal/balancer/clusterresolver/clusterresolver.go]: Updated

File: xds/internal/balancer/clusterresolver/clusterresolver.go (hunk 1 above)
Comment (aranjans) [reply xds/internal/balancer/clusterresolver/clusterresolver.go:56]: Updated

File: xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go (hunk 3 above)
Comment (purnesh42H) [reply xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go:450]: @aranjans need to change the comment too along with test name

**Review by purnesh42H — APPROVED** [review]
lgtm

**Comment by easwars** [comment]
I've started an internal discussion thread to discuss what would be the correct fix for this. Please bring your thoughts to that thread.

**Comment by purnesh42H** [comment]
> I've started an internal discussion thread to discuss what would be the correct fix for this. Please bring your thoughts to that thread.

@easwars  Should we still keep this PR open or approach will change non trivially from here?

**Comment by easwars** [comment]
I think the approach will change non-trivially from here, so we can close this for now. Thanks.

**Comment by alialobidm** [comment]
****

**Comment by alialobidm** [comment]
****

**Pull request closed without merging**